File: python/blockchain.py
'''
@author: Krisna Pranav
@brief: blockchain implementation
@copyright Copyright (c) 2025 Krisna Pranav
@date: 1-04-2025
'''

import hashlib
import time
import json

class Block:
    def __init__(self, index, previous_hash, transactions, timestamp, nonce=0):
        self.index = index
        self.previous_hash = previous_hash
        self.transactions = transactions
        self.timestamp = timestamp
        self.nonce = nonce
        self.hash = self.calculate_hash()
    
    def calculate_hash(self):
        block_string = json.dumps(self.__dict__, sort_keys=True)
        return hashlib.sha256(block_string.encode()).hexdigest()
    
    def mine_block(self, difficulty):
        while self.hash[:difficulty] != "0" * difficulty:
            self.nonce += 1
            self.hash = self.calculate_hash()
            
class Blockchain:
    def __init__(self, difficulty=2):
        self.chain = [self.create_genesis_block()]
        self.difficulty = difficulty

    def create_genesis_block(self):
        return Block(0, "0", [], time.time())

    def add_block(self, transactions):
        new_block = Block(len(self.chain), self.chain[-1].hash, transactions, time.time())
        new_block.mine_block(self.difficulty)
        self.chain.append(new_block)
        

def main():
    blockchain = Blockchain()
    blockchain.add_block(["John pays Bob 10 BTC"])
    blockchain.add_block(["Bob pays Carlos 5 BTC"])

    for block in blockchain.chain:
        print(vars(block))
        
if __name__ == "__main__":
    print("Blockchain activating...")
    main()

File: cpp/blockchain.cpp
/**
 * @file blockchain.cpp
 * @author Krisna Pranav
 * @version 1.0
 * @date 2025-04-01
 * 
 * @copyright Copyright (c) 2025 Krisna Pranav
 * 
 */

#include <iostream>
#include <vector>
#include <ctime>
#include <sstream>
#include <openssl/sha.h>

using namespace std;

class Block
{
public:
    int index;
    string previousHash;
    string transactions;
    time_t timestamp;
    int nonce;
    string hash;

    /**
     * @brief Construct a new Block object
     * 
     * @param idx 
     * @param prevHash 
     * @param txn 
     * @param time 
     */
    Block(int idx, string prevHash, string txn, time_t time) : index(idx), previousHash(prevHash), transactions(txn), timestamp(time), nonce(0)
    {
        hash = calculateHash();
    }

    string calculateHash()
    {
        stringstream ss;
        ss << index << previousHash << transactions << timestamp << nonce;
        string data = ss.str();

        unsigned char hashBuffer[SHA256_DIGEST_LENGTH];
        SHA256((unsigned char *)&data[0], data.size(), hashBuffer);

        stringstream hexStream;
        for (int i = 0; i < SHA256_DIGEST_LENGTH; ++i)
        {
            hexStream << hex << (int)hashBuffer[i];
        }
        return hexStream.str();
    }

    /**
     * @param difficulty 
     */
    void mineBlock(int difficulty)
    {
        string prefix(difficulty, '0');
        while (hash.substr(0, difficulty) != prefix)
        {
            nonce++;
            hash = calculateHash();
        }
    }
}; // class Block

class Blockchain
{
public:
    vector<Block> chain;
    int difficulty;

    /**
     * @brief Construct a new Blockchain object
     * 
     * @param diff 
     */
    Blockchain(int diff) : difficulty(diff)
    {
        chain.push_back(createGenesisBlock());
    }

    /**
     * @brief Create a Genesis Block object
     * 
     * @return Block 
     */
    Block createGenesisBlock()
    {
        return Block(0, "0", "Genesis Block", time(0));
    }

    /**
     * @param transactions 
     */
    void addBlock(string transactions)
    {
        Block newBlock(chain.size(), chain.back().hash, transactions, time(0));
        newBlock.mineBlock(difficulty);
        chain.push_back(newBlock);
    }
}; // class Blockchain

int main()
{
    Blockchain blockchain(2);
    blockchain.addBlock("John pays Bob 10 BTC");
    blockchain.addBlock("Bob pays Carlos 5 BTC");

    for (const auto &block : blockchain.chain)
    {
        cout << "Index: " << block.index << "\nHash: " << block.hash << "\nPrevious: " << block.previousHash << "\n"
             << endl;
    }
}
